### PR TITLE
Improve control flow analysis in function expressions

### LINKED ISSUE #8381: IIFEs incorrectly refine Maybe-values to Nothing
**TypeScript Version:** 

nightly (1.9.0-dev.20160429)

**Code**

```ts
let maybeNumber: number | undefined;
(function () {
    maybeNumber = 1;
})();
if (maybeNumber !== undefined) {
    maybeNumber++;
}
```

**Expected behavior:**
Everything works without errors, `maybeNumber` preserves type of `number | undefined` and refines to `number` in the `if` block.

```ts
let maybeNumber: number | undefined;
// Control flow figured that: maybeNumber === undefined
(function () {
    maybeNumber = 1;
})(); // --> oops, unknown side effects
// Revert maybeNumber to initial `number | undefined` to cover all possibilities
if (maybeNumber) {
    // Control flow: can refine to `number` again here
    maybeNumber++;
}
```

**Actual behavior:** 
Language service reports type `nothing` on `maybeNumber` after the call (even though it's `number | undefined` before the call), so statement inside of the `if` condition breaks.

### PR DESCRIPTION
**Pull request by ahejlsberg**
This PR improves control flow analysis as follows:

* Immediately invoked function expressions (IIFEs) are considered part of the surrounding control flow. Effectively, IIFEs act just like statement blocks for purposes of control flow analysis.
* Type guards for captured `const` locals are in effect within nested function expressions and arrow functions. Although it is not known when (or even whether) a function expression or arrow function will be invoked, it *is* known that once proven true a type guard for a captured `const` local will remain true (because the `const` local can't be reassigned).

Some examples:

```typescript
declare function getStringOrNumber(): string | number;

// The type guard for x remains in effect in the arrow function because x is a const local.
// Had x been declared using let or var, the type guard would not be in effect.
function f1() {
    const x = getStringOrNumber();
    if (typeof x === "string") {
        const f = () => x.length;  // Ok
    }
}

// Type guard for x remains in effect in the IIFE.
function f2() {
    let x = getStringOrNumber();
    if (typeof x === "string") {
        let n = function() {
            return x.length;  // Ok
        }();
    }
}

// Type guard for x remains in effect and assignment to y is known to
// have executed before IIFE is invoked.
function f3() {
    let x = getStringOrNumber();
    let y: number;
    if (typeof x === "string") {
        let n = (z => x.length + y + z)(y = 1);
    }
}
```

Fixes #8381.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@mhegazy @vladima @rbuckton My latest commit optimizes the critical code paths in the binder for a  bind time reduction of about 10%. Specifically, the `bindChildren` method had gotten really bloated and didn't put `getContainerFlags` to its proper full use. We'll now completely circumvent the container specific logic when the container flags are zero (which is the common case).

**Comment by ahejlsberg** [comment]
Turns out there were more opportunities for optimization in the binder. Bind time is now reduced by 25% compared to the original code. Latest gains come from completely circumventing `bindChildren` for terminal nodes (such as identifiers, literals, and other token nodes). Also, the `updateStrictMode` function was folded into `bindWorker`.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -8625,23 +8643,25 @@ namespace ts {
...
             if (isContextSensitiveFunctionOrObjectLiteralMethod(func)) {
-                const iife = getImmediatelyInvokedFunctionExpression(func);
-                if (iife) {
-                    const indexOfParameter = indexOf(func.parameters, parameter);
-                    if (iife.arguments && indexOfParameter < iife.arguments.length) {
-                        if (parameter.dotDotDotToken) {
-                            const restTypes: Type[] = [];
-                            for (let i = indexOfParameter; i < iife.arguments.length; i++) {
-                                restTypes.push(getTypeOfExpression(iife.arguments[i]));
+                if (func.kind === SyntaxKind.FunctionExpression || func.kind === SyntaxKind.ArrowFunction) {
```
Comment (mhegazy) [inline src/compiler/checker.ts]: we already check this in getImmediatelyInvokedFunctionExpression

File: src/compiler/binder.ts [hunk 2]
```diff
@@ -1098,35 +1065,64 @@ namespace ts {
...
+        function bindCallExpressionFlow(node: CallExpression) {
+            // If the target of the call expression is a function expression or arrow function we have
+            // an immediately invoked function expression (IIFE). Initialize the flowNode property to
+            // the current control flow (which includes evaluation of the IIFE arguments).
+            let expr: Expression = node.expression;
+            while (expr.kind === SyntaxKind.ParenthesizedExpression) {
+                expr = (<ParenthesizedExpression>expr).expression;
+            }
+            if (expr.kind === SyntaxKind.FunctionExpression || expr.kind === SyntaxKind.ArrowFunction) {
+                forEach(node.typeArguments, bind);
```
Comment (mhegazy) [inline src/compiler/binder.ts]: `expr.typeArgyments`  and not `node.typeArguments`.

**Comment by mhegazy** [comment]
:+1:

**Comment by mhegazy** [comment]
@vladima and @rbuckton can you take a look.

File: src/compiler/binder.ts [hunk 3]
```diff
@@ -428,115 +419,85 @@ namespace ts {
...
-                savedHasExplicitReturn = hasExplicitReturn;
-                savedCurrentFlow = currentFlow;
-                savedBreakTarget = currentBreakTarget;
-                savedContinueTarget = currentContinueTarget;
-                savedActiveLabels = activeLabels;
-
-                hasExplicitReturn = false;
-                currentFlow = { flags: FlowFlags.Start };
+            if (containerFlags & ContainerFlags.IsControlFlowContainer) {
+                const saveCurrentFlow = currentFlow;
```
Comment (rbuckton) [inline src/compiler/binder.ts]: Do you also need to save/restore the `currentTrueTarget` and `currentFalseTarget` flow labels here? I see that they are tracked in `bindConditionalExpressionFlow` and `bindPrefixUnaryExpressionFlow`, but what happens when one branch of the condition contains an IIFE?

File: src/compiler/checker.ts (hunk 1 above)
Comment (ahejlsberg) [inline src/compiler/checker.ts]: Yeah, I'll remove it.

File: src/compiler/binder.ts (hunk 3 above)
Comment (ahejlsberg) [inline src/compiler/binder.ts]: No, there's no need to save those. They're only used when a conditional operator is immediately contained in a conditional statement, and they're saved and restored at that point.

**Pull request merged**